File: examples/python/build_together.py
#!/usr/bin/env python3
"""Build Together — 3 Agents Coordinate to Build Tic-Tac-Toe

The capstone Cowchat example: a coordinator, a server-dev, and a
client-dev agent work together to "build" a tic-tac-toe game.

Demonstrates ALL coordination features: rooms, sub-rooms, messaging,
sealed-ballot voting, leader election, and decisions.

Usage:
    cargo run -p cowchat-server -- serve
    python examples/python/build_together.py
"""

import time
import uuid
from cowchat import Agent, read_api_key


def main():
    key = read_api_key()
    run_id = uuid.uuid4().hex[:6]

    # ========================================================
    print("\n=== PHASE 1: SETUP & PLANNING ===\n")
    # ========================================================

    coordinator = Agent(key, f"coordinator-{run_id}")
    server_dev = Agent(key, f"server-dev-{run_id}")
    client_dev = Agent(key, f"client-dev-{run_id}")
    print("  Connected 3 agents:")
    print(f"    coordinator: {coordinator.agent_id}")
    print(f"    server-dev:  {server_dev.agent_id}")
    print(f"    client-dev:  {client_dev.agent_id}")

    # Coordinator creates the project room
    room_name = f"tictactoe-{run_id}"
    project = coordinator.create_room(room_name, description="Tic-tac-toe game project")
    project_id = project["room_id"]
    print(f"\n  Created project room: {room_name} ({project_id})")

    # Everyone joins
    coordinator.join_room(project_id)
    server_dev.join_room(project_id)
    client_dev.join_room(project_id)
    print("  All agents joined\n")

    # Coordinator kicks things off
    coordinator.send_message(project_id,
        "Let's build a tic-tac-toe game. We need a TCP server and a client.")
    print("  coordinator: Let's build a tic-tac-toe game.")

    # Create sub-rooms for focused work
    server_room = coordinator.create_room(
        f"ttt-server-{run_id}", parent_id=project_id)
    client_room = coordinator.create_room(
        f"ttt-client-{run_id}", parent_id=project_id)
    print(f"  Created sub-rooms: ttt-server-{run_id}, ttt-client-{run_id}")

    server_dev.join_room(server_room["room_id"])
    client_dev.join_room(client_room["room_id"])
    print(f"  server-dev -> ttt-server, client-dev -> ttt-client")

    # ========================================================
    print("\n=== PHASE 2: VOTE ON PROTOCOL ===\n")
    # ========================================================

    vote = coordinator.create_vote(
        project_id,
        "Wire protocol format?",
        ["JSON", "plain text", "binary"],
        description="How should the server and client talk to each other?",
    )
    print('  Vote: "Wire protocol format?"')
    print("  Options: JSON | plain text | binary")
    print(f"  Eligible: {vote['eligible_voters']}\n")

    # Cast sealed ballots
    print("  Casting sealed ballots...")
    coordinator.cast_vote(vote["vote_id"], 0)  # JSON
    print("    coordinator -> (sealed)")
    server_dev.cast_vote(vote["vote_id"], 0)   # JSON
    print("    server-dev  -> (sealed)")
    client_dev.cast_vote(vote["vote_id"], 1)   # plain text
    print("    client-dev  -> (sealed)")

    # Wait for results
    result = coordinator.wait_for_event("vote_result")["payload"]
    print("\n  Results revealed:")
    for entry in result.get("tally", []):
        option = entry["option_text"]
        count = entry["count"]
        bar = "#" * (count * 4)
        print(f"    {option:<12} {bar} ({count})")

    ballots = result.get("ballots", [])
    if ballots:
        print()
        options = ["JSON", "plain text", "binary"]
        for ballot in ballots:
            name = ballot["agent_name"]
            choice = options[ballot["option_index"]]
            print(f"    {name:<14} voted {choice}")

    coordinator.send_message(project_id, "Vote result: JSON wins! We'll use JSON.")

    # ========================================================
    print("\n=== PHASE 3: ELECT TECH LEAD ===\n")
    # ========================================================

    coordinator.elect_leader(project_id)
    print("  Election started (2s opt-out window)...")

    # Wait for election result
    elected = server_dev.wait_for_event("leader_elected", timeout=5.0)["payload"]
    leader_id = elected["leader_id"]
    leader_name = elected["leader_name"]
    print(f"  Leader elected: {leader_name}")

    # Leader issues the protocol decision
    protocol_spec = ('JSON protocol decided: {"action":"move","pos":0-8} for moves, '
                     '{"state":"board","cells":["X","O"," ",...]} for state updates')

    agents = {coordinator.agent_id: coordinator,
              server_dev.agent_id: server_dev,
              client_dev.agent_id: client_dev}
    leader = agents[leader_id]
    leader.send_decision(project_id, protocol_spec)

    decision = coordinator.wait_for_event("decision_made")["payload"]
    content = decision.get("content", "?")
    print(f'\n  Decision by {decision.get("leader_name", "?")}: "{content[:60]}"')

    # ========================================================
    print("\n=== PHASE 4: BUILD ===\n")
    # ========================================================

    server_dev.send_message(server_room["room_id"],
        "Building game server... TCP listener, move validation, win detection")
    print("  [ttt-server] server-dev: Building game server...")

    client_dev.send_message(client_room["room_id"],
        "Building game client... board renderer, input parser, TCP connector")
    print("  [ttt-client] client-dev: Building game client...")

    # Simulate build time
    time.sleep(0.5)
    print("  Building", end="", flush=True)
    for _ in range(3):
        time.sleep(0.3)
        print(".", end="", flush=True)
    print()

    server_dev.send_message(server_room["room_id"],
        "Server done! Listening on port 3000. Supports 2-player matches.")
    print("\n  [ttt-server] server-dev: Server done! Listening on port 3000")

    client_dev.send_message(client_room["room_id"],
        "Client done! Connects to server:3000, renders board in terminal.")
    print("  [ttt-client] client-dev: Client done! Connects to server:3000")

    # Report back to main room
    server_dev.send_message(project_id, "Server component ready!")
    client_dev.send_message(project_id, "Client component ready!")
    print("\n  Both devs reported ready in main room")

    # ========================================================
    print("\n=== PHASE 5: INTEGRATION ===\n")
    # ========================================================

    coordinator.send_message(project_id,
        "All components ready. Tic-tac-toe is shipped! Great work team.")
    print("  coordinator: All components ready. Ship it!")

    # Fetch and print the full project history
    history = coordinator.get_history(project_id, limit=50)

    print(f"\n=== PROJECT HISTORY ({len(history)} messages) ===\n")
    for msg in history:
        name = msg.get("agent_name", "?")
        content = msg.get("content", "")
        metadata = msg.get("metadata") or {}
        is_decision = metadata.get("type") == "decision"
        if is_decision:
            print(f"  [DECISION] {name}: {content}")
        else:
            print(f"  {name}: {content}")

    # Clean up
    server_dev.leave_room(server_room["room_id"])
    client_dev.leave_room(client_room["room_id"])
    print("\n  Sub-rooms left")

    print("\nDone!")


if __name__ == "__main__":
    main()

File: examples/rust/build_together.rs
//! Build Together — 3 Agents Coordinate to Build Tic-Tac-Toe
//!
//! The capstone Cowchat example: a coordinator, a server-dev, and a
//! client-dev agent work together to "build" a tic-tac-toe game.
//!
//! Demonstrates ALL coordination features: rooms, sub-rooms, messaging,
//! sealed-ballot voting, leader election, and decisions.
//!
//! ```bash
//! # Start the server first:
//! cargo run -p cowchat-server -- serve
//!
//! # Then run this example:
//! cargo run -p cowchat-client --example build_together
//! ```

use cowchat_client::CowchatClient;
use cowchat_core::FrameType;
use std::path::PathBuf;
use std::time::Duration;

fn read_api_key() -> String {
    let home = std::env::var("HOME").expect("HOME not set");
    let key_path = PathBuf::from(home).join(".cowchat/auth.key");
    std::fs::read_to_string(&key_path)
        .unwrap_or_else(|_| panic!("Could not read API key from {}", key_path.display()))
        .trim()
        .to_string()
}

async fn connect(key: &str, name: &str) -> CowchatClient {
    CowchatClient::connect_tcp("127.0.0.1:9229", key, name, None, vec![])
        .await
        .unwrap_or_else(|e| panic!("Failed to connect as {name}: {e}"))
}

/// Wait for a specific event type from a subscription.
async fn wait_for(
    events: &mut tokio::sync::broadcast::Receiver<cowchat_client::Event>,
    target: FrameType,
) -> serde_json::Value {
    tokio::time::timeout(Duration::from_secs(5), async {
        loop {
            if let Ok(event) = events.recv().await {
                if event.frame.frame_type == target {
                    return event.frame.payload;
                }
            }
        }
    })
    .await
    .unwrap_or_else(|_| panic!("Timed out waiting for {target:?}"))
}

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    let key = read_api_key();

    // ========================================================
    println!("\n=== PHASE 1: SETUP & PLANNING ===\n");
    // ========================================================

    // Connect 3 agents
    let coordinator = connect(&key, "coordinator").await;
    let server_dev = connect(&key, "server-dev").await;
    let client_dev = connect(&key, "client-dev").await;
    println!("  Connected 3 agents:");
    println!("    coordinator: {}", coordinator.agent_id);
    println!("    server-dev:  {}", server_dev.agent_id);
    println!("    client-dev:  {}", client_dev.agent_id);

    // Coordinator creates the project room (unique name for re-runnability)
    let run_id = &uuid::Uuid::new_v4().to_string()[..6];
    let room_name = format!("tictactoe-{run_id}");
    let project = coordinator
        .create_room(&room_name, Some("Tic-tac-toe game project"), None)
        .await?;
    let project_id = project.room_id.clone();
    println!("\n  Created project room: {room_name} ({project_id})");

    // Everyone joins
    coordinator.join_room(&project_id).await?;
    server_dev.join_room(&project_id).await?;
    client_dev.join_room(&project_id).await?;
    println!("  All agents joined\n");

    // Coordinator kicks things off
    coordinator
        .send_message(
            &project_id,
            "Let's build a tic-tac-toe game. We need a TCP server and a client.",
            None,
            vec![],
        )
        .await?;
    println!("  coordinator: Let's build a tic-tac-toe game.");

    // Create sub-rooms for focused work
    let server_room = coordinator
        .create_room(&format!("ttt-server-{run_id}"), None, Some(&project_id))
        .await?;
    let client_room = coordinator
        .create_room(&format!("ttt-client-{run_id}"), None, Some(&project_id))
        .await?;
    println!(
        "  Created sub-rooms: ttt-server-{run_id} ({}), ttt-client-{run_id} ({})",
        server_room.room_id, client_room.room_id
    );

    // Devs join their respective sub-rooms
    server_dev.join_room(&server_room.room_id).await?;
    client_dev.join_room(&client_room.room_id).await?;
    println!("  server-dev -> ttt-server, client-dev -> ttt-client");

    // ========================================================
    println!("\n=== PHASE 2: VOTE ON PROTOCOL ===\n");
    // ========================================================

    // Subscribe to events before creating the vote
    let mut coord_events = coordinator.subscribe();

    // Coordinator creates a sealed-ballot vote
    let vote = coordinator
        .create_vote(
            &project_id,
            "Wire protocol format?",
            Some("How should the server and client talk to each other?"),
            vec![
                "JSON".to_string(),
                "plain text".to_string(),
                "binary".to_string(),
            ],
            None,
        )
        .await?;
    println!("  Vote: \"Wire protocol format?\"");
    println!("  Options: JSON | plain text | binary");
    println!("  Eligible: {}\n", vote.eligible_voters);

    // Cast sealed ballots
    println!("  Casting sealed ballots...");
    coordinator.cast_vote(&vote.vote_id, 0).await?; // JSON
    println!("    coordinator -> (sealed)");
    server_dev.cast_vote(&vote.vote_id, 0).await?; // JSON
    println!("    server-dev  -> (sealed)");
    client_dev.cast_vote(&vote.vote_id, 1).await?; // plain text
    println!("    client-dev  -> (sealed)");

    // Wait for results
    let result = wait_for(&mut coord_events, FrameType::VoteResult).await;
    println!("\n  Results revealed:");
    if let Some(tally) = result["tally"].as_array() {
        for entry in tally {
            let option = entry["option_text"].as_str().unwrap_or("?");
            let count = entry["count"].as_u64().unwrap_or(0);
            let bar = "#".repeat(count as usize * 4);
            println!("    {option:<12} {bar} ({count})");
        }
    }
    if let Some(ballots) = result["ballots"].as_array() {
        println!();
        for ballot in ballots {
            let name = ballot["agent_name"].as_str().unwrap_or("?");
            let idx = ballot["option_index"].as_u64().unwrap_or(0);
            let options = ["JSON", "plain text", "binary"];
            let choice = options.get(idx as usize).unwrap_or(&"?");
            println!("    {name:<14} voted {choice}");
        }
    }

    // Post the result to the room
    coordinator
        .send_message(
            &project_id,
            "Vote result: JSON wins! We'll use JSON.",
            None,
            vec![],
        )
        .await?;

    // ========================================================
    println!("\n=== PHASE 3: ELECT TECH LEAD ===\n");
    // ========================================================

    let mut sdev_events = server_dev.subscribe();

    coordinator.elect_leader(&project_id).await?;
    println!("  Election started (2s opt-out window)...");

    // Wait for election result
    let elected = wait_for(&mut sdev_events, FrameType::LeaderElected).await;
    let leader_id = elected["leader_id"].as_str().unwrap_or("?").to_string();
    let leader_name = elected["leader_name"].as_str().unwrap_or("?");
    println!("  Leader elected: {leader_name}");

    // Leader issues the protocol decision
    let protocol_spec = r#"JSON protocol decided: {"action":"move","pos":0-8} for moves, {"state":"board","cells":["X","O"," ",...]} for state updates"#;

    // Find which client is the leader and issue the decision
    let leader_client = if leader_id == coordinator.agent_id {
        &coordinator
    } else if leader_id == server_dev.agent_id {
        &server_dev
    } else {
        &client_dev
    };
    leader_client
        .send_decision(&project_id, protocol_spec, serde_json::json!({}))
        .await?;

    let decision = wait_for(&mut coord_events, FrameType::DecisionMade).await;
    println!(
        "\n  Decision by {}: \"{}\"",
        decision["leader_name"].as_str().unwrap_or("?"),
        &decision["content"].as_str().unwrap_or("?")[..60],
    );

    // ========================================================
    println!("\n=== PHASE 4: BUILD ===\n");
    // ========================================================

    // Server dev works in their sub-room
    server_dev
        .send_message(
            &server_room.room_id,
            "Building game server... TCP listener, move validation, win detection",
            None,
            vec![],
        )
        .await?;
    println!("  [ttt-server] server-dev: Building game server...");

    // Client dev works in their sub-room
    client_dev
        .send_message(
            &client_room.room_id,
            "Building game client... board renderer, input parser, TCP connector",
            None,
            vec![],
        )
        .await?;
    println!("  [ttt-client] client-dev: Building game client...");

    // Simulate build time
    tokio::time::sleep(Duration::from_millis(500)).await;
    print!("  Building");
    for _ in 0..3 {
        tokio::time::sleep(Duration::from_millis(300)).await;
        print!(".");
    }
    println!();

    // Server done
    server_dev
        .send_message(
            &server_room.room_id,
            "Server done! Listening on port 3000. Supports 2-player matches.",
            None,
            vec![],
        )
        .await?;
    println!("\n  [ttt-server] server-dev: Server done! Listening on port 3000");

    // Client done
    client_dev
        .send_message(
            &client_room.room_id,
            "Client done! Connects to server:3000, renders board in terminal.",
            None,
            vec![],
        )
        .await?;
    println!("  [ttt-client] client-dev: Client done! Connects to server:3000");

    // Report back to main room
    server_dev
        .send_message(&project_id, "Server component ready!", None, vec![])
        .await?;
    client_dev
        .send_message(&project_id, "Client component ready!", None, vec![])
        .await?;
    println!("\n  Both devs reported ready in main room");

    // ========================================================
    println!("\n=== PHASE 5: INTEGRATION ===\n");
    // ========================================================

    coordinator
        .send_message(
            &project_id,
            "All components ready. Tic-tac-toe is shipped! Great work team.",
            None,
            vec![],
        )
        .await?;
    println!("  coordinator: All components ready. Ship it!");

    // Fetch and print the full project history
    let history = coordinator.get_history(&project_id, 50, None).await?;

    println!("\n=== PROJECT HISTORY ({} messages) ===\n", history.len());
    for msg in &history {
        let name = &msg.agent_name;
        let content = &msg.content;
        // Check if it's a decision (has metadata with type)
        let is_decision = msg
            .metadata
            .get("type")
            .and_then(|v| v.as_str())
            .map(|t| t == "decision")
            .unwrap_or(false);

        if is_decision {
            println!("  [DECISION] {name}: {content}");
        } else {
            println!("  {name}: {content}");
        }
    }

    // Clean up — leave the sub-rooms
    server_dev.leave_room(&server_room.room_id).await?;
    client_dev.leave_room(&client_room.room_id).await?;
    coordinator.leave_room(&server_room.room_id).await.ok(); // coordinator may not have joined
    coordinator.leave_room(&client_room.room_id).await.ok();
    println!("\n  Sub-rooms left");

    println!("\nDone!");
    Ok(())
}
